experiments/sample_ssl4eo.py

```diff
@@ -0,0 +1,265 @@
+
+
+class OverlapError(Exception):
+    pass
+
+
+def sample_random_locations_rtree(
+    idx: int,
+    sampler: Any,
+    radius: float,
+    overlap_ratio: float,
+    rtree_obj: index.Index = index.Index(),
+) -> List[float]:
+    """sample new coord and check overlap --- rtree"""
+    # use rtree to avoid strong overlap
+    count = 0
+    bbox_radius = (1 - overlap_ratio) * radius / 1000
+    while count < 1:
+        new_coord = sampler.sample_point()  # (lon,lat) of top-10000 cities
+        bbox = (
+            new_coord[0] - sampler.km2deg(bbox_radius),
+            new_coord[1] - sampler.km2deg(bbox_radius),
+            new_coord[0] + sampler.km2deg(bbox_radius),
+            new_coord[1] + sampler.km2deg(bbox_radius),
+        )
+        if list(rtree_obj.intersection(bbox)):
+            continue
+        rtree_obj.insert(idx, bbox)
+        count += 1
+        center_coord = [new_coord[0], new_coord[1]]
+    return center_coord
+
+
+def fix_random_seeds(seed: int = 42) -> None:
+    np.random.seed(seed)
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--save_path", type=str, default="./data/", help="dir to save data"
+    )
+    parser.add_argument(
+        "--radius", type=float, default=1320, help="patch radius in meters"
+    )
+    parser.add_argument(
+        "--overlap_ratio",
+        type=float,
+        default=0.25,
+        help="max overlap ratio between adjacent patches",
+    )
+    parser.add_argument(
+        "--num_cities", type=int, default=10000, help="number of cities to sample"
+    )
+    parser.add_argument(
+        "--std", type=int, default=50, help="std of gaussian distribution"
+    )
+    parser.add_argument(
+        "--resume", type=str, default=None, help="resume from a previous run"
+    )
+    parser.add_argument(
+        "--indices_range",
+        type=int,
+        nargs=2,
+        default=[0, 250000],
+        help="indices to download",
+    )
+
+    args = parser.parse_args()
+
+    fix_random_seeds(seed=42)
+
+    # initialize sampler
+    sampler = GaussianSampler(num_cities=args.num_cities, std=args.std)
+
+    # if resume
+    ext_coords = {}
+    if args.resume:
+        ext_path = args.resume
+        with open(ext_path) as csv_file:
+            reader = csv.reader(csv_file)
+            for row in reader:
+                key = row[0]
+                val1 = float(row[1])
+                val2 = float(row[2])
+                ext_coords[key] = (val1, val2)  # lon, lat
+    else:
+        ext_path = os.path.join(args.save_path, "sampled_locations.csv")
+
+    rtree_coords = index.Index()
+    bbox_radius = (
+        (1 - args.overlap_ratio) * args.radius / 1000
+    )  # allow little overlap between adjacent patches
+    if args.resume:
+        print("Load existing locations.")
+        for i, key in enumerate(tqdm(ext_coords.keys())):
+            c = ext_coords[key]
+            rtree_coords.insert(
```

Comment:
This is still way more complicated than it needs to be. You're inserting into an R-tree two times for every point you sample, here and on line 165. The design I proposed [here](https://github.com/microsoft/torchgeo/pull/1213#pullrequestreview-1367905066) is only 9 lines of code and should do the same thing but twice as fast. Every single insert is $\mathcal{O}(n)$, and we insert $n$ times. So the overall complexity of this sampling script is $\mathcal{O}(n^2)$, so doubling that is bad.

I think a lot of the fancy abstractions in this script (classes and a multitude of functions) are actually making it much slower. I think you can delete 90% of it and just use a single while loop:
```python
with open('locations.csv', 'w', newline='') as f:
    w = csv.writer(f)
    idx = Index()
    counter = 0
    while counter < total_num_patches:
        bbox = choose_random_bbox()  # this might be worth making a separate function for
        if list(idx.intersection(bbox)):
            continue
        idx.insert(counter, bbox)
        w.writerow(bbox)
        counter += 1
```
This is what I mean when I say the entire script could be ~30 lines of code.